### PR TITLE
Error handling refactor

### PR DESCRIPTION
**Pull request by neekolas**
## tl;dr

- Refactors `MessageProcessingError` into `GroupMessageProcessingError` that is more narrowly scoped
- Uses `GroupError` to handle errors when processing Welcomes
- Make `process_for_id` take a generic error type to support both
- Trim down unused error variants
- Make `RetryableError` explicit and don't use `_ => false`

### CODE REVIEW

**Comment by neekolas** [comment]
* **#1280** <a href="https://app.graphite.dev/github/pr/xmtp/libxmtp/1280?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.dev/github/pr/xmtp/libxmtp/1280?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `main`

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: xmtp_mls/src/groups/mls_sync.rs [hunk 1]
```diff
@@ -78,6 +78,83 @@ use xmtp_proto::xmtp::mls::message_contents::plaintext_envelope::v2::MessageType
...
+    ClearPendingCommit(#[from] sql_key_store::SqlKeyStoreError),
+    #[error("Serialization/Deserialization Error {0}")]
+    Serde(#[from] serde_json::Error),
+    #[error("intent is missing staged_commit field")]
+    IntentMissingStagedCommit,
+    #[error("encode proto: {0}")]
+    EncodeProto(#[from] prost::EncodeError),
+    #[error("proto decode error: {0}")]
+    DecodeProto(#[from] prost::DecodeError),
+    #[error(transparent)]
```
Comment (mchenani) [inline xmtp_mls/src/groups/mls_sync.rs:120]: question: in previous code we have this comment:
`    #[error("intent error: {0}")]
    Intent(#[from] IntentError),`
    is this change intentional?

File: xmtp_mls/src/groups/mod.rs [hunk 2]
```diff
@@ -216,9 +215,41 @@ impl RetryableError for GroupError {
...
+            | Self::DmGroupMetadataForbidden
+            | Self::Signature(_)
+            | Self::LeafNodeError(_)
+            | Self::NoPSKSupport
+            | Self::MissingPendingCommit
+            | Self::InvalidPermissionUpdate
+            | Self::AddressValidation(_)
+            | Self::InvalidPublicKeys(_)
+            | Self::CredentialError(_)
+            | Self::EncodeError(_) => false, // _ => false,
```
Comment (mchenani) [inline xmtp_mls/src/groups/mod.rs:252]: do we need this comment anymore?

Comment (neekolas) [reply xmtp_mls/src/groups/mod.rs:252]: Good catch, will remove

**Review by mchenani — APPROVED** [review]
Nice cleanup🧹🙌
